### PR TITLE
Fix output serialization for cairo 1

### PR DESCRIPTION
**Pull request by fmoletta**
This PR reverts #1630 and re-implements it

- Adds handling of Spans & Dictionaries to the existing serialization loop, using a similar serialization format to the current Array serialization
- Fixes bug reported by #1644

### CODE REVIEW

**Comment by raphaelDkhn** [comment]
Thanks @juanbono and @fmoletta  this PR as well as #1630 successfully fix the issue with Nullable Felt252Dict printing. However, a problem arises when Nullable Felt252Dict types are embeded within a struct. The printing does not behave as expected. Let's take the following example:

```rust
struct NullableVec<T> {
    items: Felt252Dict<Nullable<T>>,
    len: usize,
}

fn main() -> NullableVec<u32> {
    let mut d: Felt252Dict<Nullable<u32>> = Default::default();

    // Populate the dictionary
    d.insert(0, nullable_from_box(BoxTrait::new(10)));
    d.insert(1, nullable_from_box(BoxTrait::new(20)));
    d.insert(2, nullable_from_box(BoxTrait::new(30)));

    // Return a populated NullableVec
    NullableVec {
        items: d,
        len: 3,
    }
}

```

When executing the code with `cargo run programs/null_vec.sierra --layout all_cairo --print_output`, the expected output should display the values as follow`{0:10 1:20 2:30} 3`. 
However, the actual output is `{0:8:0 1:8:1 2:8:2} 3`.

**Comment by fmoletta** [comment]
> Thanks @juanbono and @fmoletta this PR as well as #1630 successfully fix the issue with Nullable Felt252Dict printing. However, a problem arises when Nullable Felt252Dict types are embeded within a struct. The printing does not behave as expected. Let's take the following example:
> 
> ```rust
> struct NullableVec<T> {
>     items: Felt252Dict<Nullable<T>>,
>     len: usize,
> }
> 
> fn main() -> NullableVec<u32> {
>     let mut d: Felt252Dict<Nullable<u32>> = Default::default();
> 
>     // Populate the dictionary
>     d.insert(0, nullable_from_box(BoxTrait::new(10)));
>     d.insert(1, nullable_from_box(BoxTrait::new(20)));
>     d.insert(2, nullable_from_box(BoxTrait::new(30)));
> 
>     // Return NullableVec
>     NullableVec {
>         items: d,
>         len: 3,
>     }
> }
> ```
> 
> When executing the code with `cargo run programs/null_vec.sierra --layout all_cairo --print_output`, the expected output should display the values as follow`{0:10 1:20 2:30} 3`. However, the actual output is `{0:8:0 1:8:1 2:8:2} 3`.

Hello! Thanks for pointing this out!
I recently pushed some changes to dereference referenced values upon deserialization which should fix this issue

**Review by pefontana — APPROVED** [review]

File: cairo1-run/src/main.rs [hunk 1]
```diff
@@ -320,6 +323,116 @@ fn main() -> Result<(), Error> {
     }
 }
 
+// Serializes the return values in a user-friendly format
+// Displays Arrays using brackets ([]) and Dictionaries using ({})
+// Recursively dereferences referenced values (such as Span & Box)
```
Comment (azteca1998) [inline cairo1-run/src/main.rs:328]: If this is intended to be the documentation, then it should be prefixed with the triple slash `///`.

**Review by azteca1998 — APPROVED** [review]

**Review by igaray — APPROVED** [review]

File: cairo1-run/src/main.rs (hunk 1 above)
Comment (fmoletta) [reply cairo1-run/src/main.rs:328]: Good call!

**Pull request merged**